src/os/bluestore/BlueStore.cc

```diff
@@ -596,6 +617,121 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
   return out << ")";
 }
 
+namespace {
+
+/*
+ * Due to a bug in key string encoding (see a comment for append_escaped)
+ * the KeyValueDB iterator does not lexicographically sort the same
+ * way that ghobject_t does: objects with the same hash may have wrong order.
+ *
+ * This is the iterator wrapper that fixes the keys order.
+ */
+
+class CollectionListIterator {
+public:
+  CollectionListIterator(const KeyValueDB::Iterator &it)
+    : m_it(it), m_chunk_iter(m_chunk.begin()) {
+  }
+
+  bool valid() const {
+    return m_key != nullptr;
+  }
+
+  const ghobject_t &oid() const {
+    return *m_oid;
+  }
+
+  const std::string &key() const {
+    return *m_key;
+  }
+
+  void lower_bound(const std::string &key) {
+    ghobject_t oid;
+    int r = get_key_object(key, &oid);
+    ceph_assert(r != -1);
+
+    std::string pkey;
+    _key_encode_prefix(oid, &pkey);
+
+    m_it->lower_bound(pkey);
+    if (!get_next_chunk()) {
+      return;
+    }
+
+    auto iter = m_chunk.find(oid);
```

Comment:
shouldn't this be lower_bound not find? to proceed with the next entry after oid if the latter is missing.

Reply:
We have two possible cases:

1) The `key` exists in the db. In this case, after `m_it->lower_bound(pkey)` above, we load a sorted chunk of key with the same hash as the `key` has (and the `key` is among them).  And `find` will set the iterator to the `key` (as expected).

2) The `key` does not exist in the db. In this case, after `m_it->lower_bound(pkey)` above, we load a sorted chunk of key with the hash that is the next after the `key`'s hash. And in this case we need to have the iterator to set to the first element of this chunk.

Reply:
what's about the case when there is no exact key match but there are multiple different oids with the same hash as for the original input key. Hence you might need to find proper chunk, i.e. chunk next(!) to the input one not the first one

Reply:
hence you need to replace find with lower_bound AND proceed with the next chunk if this lower_bound returns end() iterator

Reply:
Indeed. Thanks. Updated.